Remove as many generic error() messages from base as possible

jakebolewski opened the pull request:
- throw more specific Exception types
- make error messages more consistent
- give more context for the error when possible
- update tests

jakebolewski committed:
Try and remove as many generic error() messages from base as possible

 * throw more specific Exception types
 * make error messages more consistent
 * give more context for the error when possible
 * update tests

GlenHertz commented on base/socket.jl:
```diff
@@ -21,9 +21,9 @@ end
 function IPv4(host::Integer)
     if host < 0
-        error("IP address must not be negative")
+        throw(ArgumentError("IPv4 address must not be ≥ 0"))
```
Isn't the sense on this statement wrong?  The IPv4 must be positive.

IainNZ replied:
Do these inequality symbols display correctly on Windows?

jakebolewski replied:
@GlenHertz thanks for reading through.  "must be positive" I think is subjectively better than "not negative", I'll push an update.

@IainNZ I have no idea.

GlenHertz commented on base/socket.jl:
```diff
@@ -63,11 +63,11 @@ end
 function IPv6(host::Integer)
     if host < 0
-        error("IP address must not be negative")
+        throw(ArgumentError("IPv6 address must not be ≥ 0"))
```
Wrong sense here too?

ViralBShah commented:
This is great, and much needed.

jakebolewski commented:
I've updated with @GlenHertz's suggestions and did another pass over what I pushed last night.  I also updated some of the errors in the `LinearAlgebra` module.

IainNZ commented:
Really nice work @jakebolewski !

staticfloat commented:
:+1:
